docs/proposals/vip-4.md

```diff
@@ -0,0 +1,273 @@
+  it.
+
+* `venice.store.lifecycle.hooks.wait.interval.seconds`: The time to wait before re-invoking a hook which returned `WAIT` 
+  (default: 60 seconds).
+
+* `venice.store.lifecycle.hooks.configs.<arbitrary>`: Any number of configs to be passed (after clipping everything
+  before the `<arbitrary>` part) into the hooks constructor.
+
+In addition, the store config will get a new `Map<String, String>` of store-level config overrides. Those configs are
+"[stringly-typed](https://wiki.c2.com/?StringlyTyped)", rather than strongly-typed, since we are not aware of the 
+configs needed by each hook at compile-time, and we therefore cannot shape the definition of the store config schema 
+accordingly. This issue is mitigated via the `validateHookConfig`, which can be used to prevent invalid configs from
+entering the system.
+
+### Metrics
+
+There needs to be new metrics to monitor hook health. Each new metric will be per function and per registered hooks 
+class (i.e. 13 functions per class if we implement all of them, or less if we cut the scope). For each class/function
+hook, there will be:
+
+* the occurrence rate of:
+  * hook invocations
+  * each hook return signal (for the functions that return something other than `void`)
+  * failures (exceptions)
+  * timeouts
+* the time spend waiting in queue before being executed (which will be useful to determine if the thread pool count is 
+  under-provisioned)
+
+### Design Recommendation
+
+Running hooks in the parent controller is probably most straightforward. The only issue is the inability to support the
+`postStoreVersionLeaderReplication` hook, but that one is not critical and could be left out of scope.
+
+Concretely, choosing the parent controller path would work like this:
+
+1. The parent controller would invoke the hooks for `preStartOfStoreVersionIngestionForDaVinci` and for
+   `preStoreVersionCreation` for each of the regions. If all hooks respond with `PROCEED` then it's essentially a normal 
+   push, otherwise it would configure the `AddVersion` command sent to the admin channel with the appropriate inclusions
+   in `targetedRegions` (depending on the result of `preStoreVersionCreation`) and in a new field for controlling the
+   DVC ingestion (depending on the result of `preStartOfStoreVersionIngestionForDaVinci`) which the child controller 
+   would honor by holding off on writing to the Meta Store.
+
+2. The parent controller would create the new store-version with `versionSwapDeferred = true`.
+
+3. When the parent controller sees that a region has completed ingestion, it would invoke the `preStoreVersionSwap`
+   hook for that region, and if the hook responds with `PROCEED`, then it would send an admin command targeted for that
+   region to swap the current version.
+
+4. If at any stage the hooks respond with `ROLLBACK` then the parent controller would send more admin channel commands
```

Comment:
When we rollback a store version, we don't rollback the registered schema, right?

Reply:
Correct. And the schema hook does not provide the ability to rollback.

Reply:
Got it.